### PR TITLE
Improvements - Was "Communication with CentralReport Online"

### PR DESCRIPTION
**Pull request by Guimove**
## Update 03/03/2014

CentralReport Online is not planned anymore. Another project will be born in few days and will be announced on the [CR blog](http://blog.centralreport.net).
As this pull request adds useful features and many improvements in the core of this project, this branch has been merged into develop.

---
### Original pull request:

This pull request adds the communication with CentralReport Online (centralreport.net).
CentralReport Online is an online tool to follow all your hosts, in a single place. This part is currently in development and will be enhanced in the next weeks.
- [x] Add "requests" library
- [x] Create the "webservice" module
- [x] Create webservices entities
- [x] Clean and code optimization 

Remains to be done:
- [x] Clean values from the collector ('\n' in some variables, or too many spaces in others)
- [x] Clean useless entities ("Full", "Registration")
- [x] Compact the JSON output (on a single line, without useless spaces)
- [x] Cast dates and float to long (like disk spaces or memory data)
- [x] Update the thread "Checks" to send data to the Online part
- [x] Do more tests with the JSON generation
- [x] Add the "Agent name" and "Agent version" in the host json.

_Updated tasks (miniche, 02/04/2013)_
_Updated tasks (miniche, 29/05/2013)_
_Updated title and description (miniche, 30/07/2013)_

### CODE REVIEW

**Comment by Ninir** [comment]
Please merge develop as it conflicts.

**Comment by miniche** [comment]
Remains to be done:
- Clean values from the collector ('\n' in some variables, or too many spaces in others)
- Clean useless entities ("Full", "Registration" and "Disks")
- Compact the JSON output (on a single line, without useless spaces)
- Cast dates and float to long (like disk spaces or memory data)
- Update the thread "Checks" to send data to the Online part
- Do more tests with the JSON generation
- Add the "Agent name" and "Agent version" in the host json.

File: centralreport/cr/collectors.py [hunk 1]
```diff
@@ -419,29 +425,54 @@ def get_disks(self):
...
-                # Using new check entity
                 check_disk = checks.Disk()
-                check_disk.date = datetime.datetime.now()
-                check_disk.name = line_split[0]
-                check_disk.unix_name = line_split[0]
-                check_disk.size = disk_total
-                check_disk.used = disk_used
-                check_disk.free = disk_free
+                check_disk.name = disk_name.replace('/dev/', '')
+                check_disk.unix_name = disk_name
```
Comment (Ninir) [inline centralreport/cr/collectors.py]: I'm not sure about `unix_name` here. As we have checks for UNIX only this does not matter. However, we'll probably have Windows Checks soon... Is there a specific reason?

File: centralreport/cr/entities/webservices.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+
+"""
+    CentralReport - Webservices module
+        Contains all entities used with webservices.
+
+    https://github.com/miniche/CentralReport/
```
Comment (Ninir) [inline centralreport/cr/entities/webservices.py]: Mmmm... let's say no :D

File: centralreport/cr/errors.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+
+"""
+    CentralReport - Errors module
+        Contains all definitions of the CentralReport custom errors
```
Comment (Ninir) [inline centralreport/cr/errors.py]: If errors belongs to CentralReport, then they are implicitly custom. This, I would remove the "custom" word

File: centralreport/cr/online.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
+# -*- coding: utf-8 -*-
+
+"""
+    CentralReport - Online module
+        Contains all classes and functions used for the interaction with the Online server
```
Comment (Ninir) [inline centralreport/cr/online.py]: `Contains all classes and functions interacting with the Online server`

File: centralreport/cr/online.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+    log.log_debug('Generating the JSON template...')
+    template = jinja_env.get_template('host_registration.json')
+
+    json_vars = dict()
+    json_vars['uuid'] = data.host_info.uuid
+    json_vars['hostname'] = data.host_info.hostname
+    json_vars['displayName'] = data.host_info.hostname
+    json_vars['model'] = data.host_info.model
+    json_vars['cpuCount'] = data.host_info.cpu_count
```
Comment (Ninir) [inline centralreport/cr/online.py]: What about the cpuModel as in the last file you added on DropBox?

File: centralreport/cr/online.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    json_vars['os'] = data.host_info.os_name
+    json_vars['osVersion'] = data.host_info.os_version
+    json_vars['kernel'] = data.host_info.kernel_name
+    json_vars['kernelVersion'] = data.host_info.kernel_version
+    json_vars['architecture'] = data.host_info.architecture
+    json_vars['agent'] = 'TODO'
+    json_vars['agentVersion'] = 'TODO'
```
Comment (Ninir) [inline centralreport/cr/online.py]: You are not forced to add them.. :D

File: centralreport/cr/online.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    ws_checks = web.send_data(web.METHOD_POST, route_checks_add, check_json)
+
+    if ws_checks.code == 400:
+        raise OnlineError(1, 'Data refused by the remote server: bad data')
+    elif ws_checks.code == 404:
+        raise OnlineError(2, 'Wrong KEY or UUID!')
+    elif ws_checks.code == 409:
+        raise OnlineError(3, 'Check already sent!')
+    elif ws_checks.code != 200:
+        raise OnlineError(4, 'Unkown answer code %s!' % ws_checks.code)
```
Comment (Ninir) [inline centralreport/cr/online.py]: Unknown*

Comment (Ninir) [inline centralreport/cr/online.py]: Btw, the server will return a 201 code as this is a new ressource added

File: centralreport/cr/static/online/send_checks.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            "user": {{ cpu_user }}
+        },
+        "memory": {
+            "total": {{ memory_total }},
+            "active": {{ memory_active }},
+            "inactive": {{ memory_inactive }},
+            "resident": {{ memory_resident }},
+            "free": {{ memory_free }},
+            "swapSize": {{ swap_size }},
+            "swapUsed": {{ swap_used }},
```
Comment (Ninir) [inline centralreport/cr/static/online/send_checks.json]: usedSwap

File: centralreport/cr/static/online/send_checks.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            "swapFree": {{ swap_free }}
```
Comment (Ninir) [inline centralreport/cr/static/online/send_checks.json]: freeSwap

File: centralreport/cr/threads.py [hunk 10]
```diff
@@ -59,49 +50,53 @@ def __init__(self):
...
-                # Checking disks informations
-                if convert_text_to_bool(Config.get_config_value('Checks', 'enable_disks_check')):
-                    log.log_debug('Doing a disk check...')
-                    Checks.last_check_disk = self.MyCollector.get_disks()
+        log.log_debug('---- New check -----')
+        check_entity = checks.Check()
 
-                # Updating last check date...
+        # Checking CPU
+        log.log_debug('Doing a CPU check...')
```
Comment (Ninir) [inline centralreport/cr/threads.py]: Performing is more adapted ;)

File: centralreport/cr/static/online/send_checks.json [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            "idle": {{ cpu_idle }},
+            "system": {{ cpu_system }},
+            "user": {{ cpu_user }}
+        },
+        "memory": {
+            "total": {{ memory_total }},
+            "active": {{ memory_active }},
+            "inactive": {{ memory_inactive }},
+            "resident": {{ memory_resident }},
+            "free": {{ memory_free }},
```
Comment (Ninir) [inline centralreport/cr/static/online/send_checks.json:20]: What do you think of using `freeMemory` (as for the tmpl_var)

File: centralreport/cr/online.py (hunk 5 above)
Comment (miniche) [inline centralreport/cr/online.py]: It's an oversight. I forgot to add this item in the documentation. (http://docs.centralreport.net/cro/api/users/hosts#postapiuserskeyhosts). Will be fixed tonight.

File: centralreport/cr/online.py (hunk 6 above)
Comment (miniche) [inline centralreport/cr/online.py]: It's just to avoid forgetting it :D

File: centralreport/cr/online.py (hunk 7 above)
Comment (miniche) [inline centralreport/cr/online.py]: Mmh, ok. We must update the documentation in this case (http://docs.centralreport.net/cro/api/users/hosts/checks).

File: centralreport/cr/collectors.py (hunk 1 above)
Comment (miniche) [inline centralreport/cr/collectors.py]: No specific reason, except that the Python version will be only designed for Unix-like systems.
On Windows, this name doesn't exist and replaced by a drive letter (like 'C:' or 'Z:', etc.).
If you want to standardize this, no problem, but we have to found a generic name for this attribute.

File: centralreport/cr/static/online/send_checks.json (hunk 11 above)
Comment (miniche) [inline centralreport/cr/static/online/send_checks.json:20]: Are you talking about the key (`"free"`) or about the variable name (`{{ memory_free }}`)?

Comment (Ninir) [inline centralreport/cr/static/online/send_checks.json:20]: The key :)

File: centralreport/cr/online.py (hunk 6 above)
Comment (Ninir) [inline centralreport/cr/online.py]: You are not forced to remember them too... :D

File: centralreport/cr/online.py (hunk 5 above)
Comment (Ninir) [inline centralreport/cr/online.py]: Hope so otherwise you'll be thrown into abysses with piranhas :D

File: centralreport/cr/collectors.py (hunk 1 above)
Comment (Ninir) [inline centralreport/cr/collectors.py]: disk_name seems good :P

File: centralreport/cr/online.py (hunk 5 above)
Comment (miniche) [inline centralreport/cr/online.py]: :fearful:

File: centralreport/cr/collectors.py (hunk 1 above)
Comment (miniche) [inline centralreport/cr/collectors.py]: In this case, we must rename the attribute "name" in the same class.
We could rename "name" to "display_name" and "unix_name" to "name".

Example, for OS X:

```
name = 'disk0s2'
display_name = 'Macintosh SSD'
```

(currently, it's `name = 'Macintosh SSD'` and `unix_name = 'disk0s2'`)

**Comment by miniche** [comment]
Updated pull request description.

File: centralreport/cr/collectors.py (hunk 1 above)
Comment (Ninir) [inline centralreport/cr/collectors.py]: displayName seems fine to me too :)

Comment (miniche) [inline centralreport/cr/collectors.py]: Great :)

**Comment by Guimove** [comment]
Can you merge develop in web services ?

**Comment by miniche** [comment]
Sure! Will be done this evening.

**Comment by miniche** [comment]
The official documentation (http://docs.centralreport.net/cro/api) has been updated according to c11f02c

**Comment by miniche** [comment]
Hey guys! Great news... I think it's time to review this pull request :D

File: NEWS [hunk 12]
```diff
@@ -8,6 +8,10 @@
 Version Alpha 0.3.0 (Not released yet)
     Currently in development
 
+    - Feature #52: Communication with CentralReport Online
+        * CentralReport Online is an online tool to follow all your hosts, in
```
Comment (Ninir) [inline NEWS]: `Follow` has a different meaning. What do you think of using `to check all your hosts` or `to get information of all your hosts`?

File: centralreport/cr/online.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+from cr import log
+from cr.utils import text
+from cr.utils import web
+from cr.tools import Config
+
+# Main routes. Use the HATEOAS concept (http://en.wikipedia.org/wiki/HATEOAS).
+# They are empty when CR is starting, they will be gotten dynamically from the server.
+# TODO: Use HATEOAS for "route_host_check" when the CentralReport Online API will be updated
+route_user_check = 'http://centralreport.net/api/users/%key%'
+route_host_check = 'http://centralreport.net/api/users/%key%/hosts/%uuid%'
```
Comment (Ninir) [inline centralreport/cr/online.py]: Why are you using this? Are you waiting for the `host_search` HATEOAS route?

Comment (miniche) [inline centralreport/cr/online.py]: Yes! I have added this route to perform first tests. I have added a TODO (line 25) too, to remove this URL when the `host_search` will be available.

File: NEWS (hunk 12 above)
Comment (miniche) [inline NEWS]: True.. I will fix it tonight ;-)

**Comment by Ninir** [comment]
Build failed :(

**Comment by miniche** [comment]
@Ninir Fixed since fed995d :)

**Comment by miniche** [comment]
This pull request will be cleaned in order to be merged in the develop branch, because this branch fixes and adds important stuff.
All related code to CentralReport Online will be removed, as this project will be reborn.

You can follow [@Central_Report](https://twitter.com/Central_Report/) on Twitter, we will tweet details about the future :)

**Pull request merged**